packages/components-kit/src/components/ComponentForm/ComponentForm.component.js

```diff
@@ -0,0 +1,191 @@
+import kit from 'component-kit.js';
+
+export const DEFAULT_STATE = new Map({
+	dirty: false,
+});
+
+export class TCompForm extends React.Component {
+	static displayName = 'ComponentForm';
+	static propTypes = {
+		...cmfConnect.propTypes,
+	};
+	static defaultProps = {
+		customTriggers: () => {},
+	};
+
+	static ON_CHANGE = 'TCOMP_FORM_CHANGE';
+	static ON_TRIGGER = 'TCOMP_FORM_TRIGGER';
+	static ON_SUBMIT = 'TCOMP_FORM_SUBMIT';
+	static ON_DEFINITION_URL_CHANGED = 'TCOMP_FORM_DEFINITION_URL_CHANGE';
+
+	constructor(props) {
+		super(props);
+		this.state = {};
+		this.onTrigger = this.onTrigger.bind(this);
+		this.onChange = this.onChange.bind(this);
+		this.onSubmit = this.onSubmit.bind(this);
+		this.getUISpec = this.getUISpec.bind(this);
+		this.setupTrigger = this.setupTrigger.bind(this);
+		this.getTriggers = this.getTriggers.bind(this);
+		this.setupTrigger(props);
+	}
+
+	componentDidUpdate(props) {
+		if (
+			props.triggerURL !== this.props.triggerURL ||
+			props.customTriggers !== this.props.customTriggers
+		) {
+			this.trigger = this.setupTrigger(this.props);
+		}
+		if (this.props.definitionURL !== props.definitionURL) {
+			this.props.dispatch({
+				type: TCompForm.ON_DEFINITION_URL_CHANGED,
+				...this.props,
+			});
+		}
+	}
+
+	onChange(event, data) {
+		event.persist();
+		if (!this.props.state.get('dirty')) {
+			this.props.setState({ dirty: true });
+		}
+		const properties = data.properties;
+		if (data.schema.titleMap && data.value) {
+			// Here we add a field side by side with the value
+			// to keep the title associated to the value
+			const info = data.schema.titleMap.find(titleMap => titleMap.value === data.value);
+			let currentProp = properties;
+			let currentKey;
+			data.schema.key.forEach((key, index) => {
+				if (index !== data.schema.key.length - 1) {
+					currentProp = currentProp[key];
+				} else {
+					currentKey = key;
+				}
+			});
+			currentProp[`$${currentKey}_name`] = info.name;
+		}
+		this.setState({ properties });
+		if (this.props.dispatchOnChange) {
+			this.props.dispatch({
+				type: TCompForm.ON_CHANGE,
+				event: {
+					type: 'onChange',
+					component: 'TCompForm',
+					props: this.props,
+					state: this.state,
+					source: event,
+				},
+				data,
+				properties,
+				uiSpec: this.getUISpec(),
+			});
+		}
+	}
+
+	onTrigger(event, payload) {
+		return this.trigger(event, payload).then(data => {
+			if (data.properties) {
+				this.setState({ properties: data.properties });
+			}
+			if (data.errors || data.jsonSchema || data.uiSchema) {
+				this.props.setState(data);
+			}
+			this.props.dispatch({
+				type: TCompForm.ON_TRIGGER,
+				event: {
+					type: 'onTrigger',
+					component: ComponentForm,
+					componentId: this.props.componentId,
+					props: this.props,
+					state: this.state,
+				},
+				data,
+				uiSpec: this.getUISpec(),
+			});
+		});
+	}
+
+	onSubmit(event, data) {
+		event.persist();
```

Comment:
what happen if a submit is trigerred before a running trigger is not yet resolved.

Reply:
Ok I talked to @jmfrancois, seems that even with the `event.persist()`, we are not sure to have the same event during saga execution.
The plan is to extract useful infos from event to add it in action to dispatch. This point has been added to the component form TODO list #1525 (let's try not to make this PR bigger, but be sure that the list is all the points that must be fixed to be usable).